### PR TITLE
sql/catalog: generalize the lease manager to lease all descriptors

### PR DESCRIPTION
**Pull request by thoszhang**
This PR extends the lease manager to lease `catalog.Descriptors` in general. Details are in individual commits.

### CODE REVIEW

File: pkg/sql/catalog/descs/collection.go [hunk 1]
```diff
@@ -380,6 +380,15 @@ func (tc *Collection) GetTableVersion(
 		// know how to deal with, so propagate the error.
 		return nil, err
 	}
+	table, ok := desc.(*sqlbase.ImmutableTableDescriptor)
+	if !ok {
+		// TODO (lucy): Is this the right error? I think it should be possible at
```
Comment (rohany) [inline pkg/sql/catalog/descs/collection.go:385]: I think we have some errors like "ID is not a table" that we use in these cases.

Comment (rohany) [reply pkg/sql/catalog/descs/collection.go:385]: But table does not exist when you find a type is a fine error

File: pkg/sql/sqlbase/schema_desc.go [hunk 2]
```diff
@@ -106,3 +126,20 @@ func (desc *ImmutableSchemaDescriptor) DescriptorProto() *Descriptor {
 
 // NameResolutionResult implements the ObjectDescriptor interface.
 func (desc *ImmutableSchemaDescriptor) NameResolutionResult() {}
+
+// MaybeIncrementVersion implements the MutableDescriptor interface.
+func (desc MutableSchemaDescriptor) MaybeIncrementVersion() {
```
Comment (rohany) [inline pkg/sql/sqlbase/schema_desc.go:131]: This should be a pointer

File: pkg/sql/sqlbase/type_desc.go [hunk 3]
```diff
@@ -170,6 +171,23 @@ func (desc *TypeDescriptor) TypeName() string {
 	return "type"
 }
 
+// MaybeIncrementVersion implements the MutableDescriptor interface.
+func (desc MutableTypeDescriptor) MaybeIncrementVersion() {
```
Comment (rohany) [inline pkg/sql/sqlbase/type_desc.go:175]: pointer as well?

**Review by rohany — COMMENTED** [review]
Nice work! Left a few comments, but this isn't code I'm familiar with. I'll leave the review to others.

**Review by ajwerner — APPROVED** [review]
Nice :lgtm: mod @rohany's pointer points and nits

> Reviewed 1 of 4 files at r2, 9 of 10 files at r10, 14 of 15 files at r11, 1 of 25 files at r12, 1 of 15 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-:-MBAZjNA1JDgBtyZ8pea:baylnbq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @rohany)

___
*[pkg/sql/crdb_internal.go, line 420 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-MBAOJrF2zhWkRPEWwIV:-MBAOJrGEG6eZ453Kz_i:b8h3pb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/crdb_internal.go#L420)):*
> ```Go
> 		nodeID := tree.NewDInt(tree.DInt(int64(p.execCfg.NodeID.Get())))
> 		p.LeaseMgr().VisitLeases(func(desc catalog.Descriptor, dropped bool, _ int, expiration tree.DTimestamp) (wantMore bool) {
> 			desc = desc.(*ImmutableTableDescriptor)
> ```

Do we need to do this type assertion? Which method is not on `Descriptor`? just `GetParentID()`, no?  At least add a TODO here?
___
*[pkg/sql/catalog/catalog.go, line 21 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-MBA_93v1glEmCcXHGt9:-MBA_93wEE0n_y3UtljE:bhu98ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/48a66b8ec7ff645d32f52ce32c944449764b0f72/pkg/sql/catalog/catalog.go#L21)):*
> ```Go
> type Descriptor = sqlbase.DescriptorInterface
> 
> type MutableDescriptor interface {
> ```

This needs a comment.
___
*[pkg/sql/catalog/descs/collection.go, line 385 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-M9uxQuqlRXZjSnm_WXz-r11-385:-MBAQBt13L4G22asKtLU:bx502p6) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/catalog/descs/collection.go#L385)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

But table does not exist when you find a type is a fine error
</blockquote></details>

WrongObjectType? 

https://github.com/cockroachdb/cockroach/blob/d5229de2e4027a200e2b235fa794d7dbf74a646a/pkg/sql/catalog/catalogkv/catalogkv.go#L285 
___
*[pkg/sql/sqlbase/structured.go, line 1700 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-MBAP6-T2RS-_GtCtJa8:-MBAP6-T2RS-_GtCtJa9:b-c0kvwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/sqlbase/structured.go#L1700)):*
> ```Go
> 	}
> 	desc.Version++
> 	desc.ModificationTime = hlc.Timestamp{}
> ```

:grimacing: I just realized I had not successfully merged this earlier today. https://github.com/cockroachdb/cockroach/pull/42655 I was planning on bors'ing this, it'll be a minor rebase.

**Review by thoszhang — COMMENTED** [review]
Still trying to figure out this opaque test failure that produced no useful CI output.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-:-MBAyeuoBt31BHroYEZb:b7dy2v7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @rohany)

___
*[pkg/sql/crdb_internal.go, line 420 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-MBAOJrF2zhWkRPEWwIV:-MBAxp70EMRsd3nOVPxK:b-2lhr6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/crdb_internal.go#L420)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Do we need to do this type assertion? Which method is not on `Descriptor`? just `GetParentID()`, no?  At least add a TODO here?
</blockquote></details>

Yeah, this shouldn't be here, I removed it.
___
*[pkg/sql/catalog/catalog.go, line 21 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-MBA_93v1glEmCcXHGt9:-MBAxtV6B0ViEBwCbzHB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/48a66b8ec7ff645d32f52ce32c944449764b0f72/pkg/sql/catalog/catalog.go#L21)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This needs a comment.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/collection.go, line 385 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-M9uxQuqlRXZjSnm_WXz-r11-385:-MBAxuN71TPvOuIatW9P:b-5uzlov) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/catalog/descs/collection.go#L385)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

WrongObjectType? 

https://github.com/cockroachdb/cockroach/blob/d5229de2e4027a200e2b235fa794d7dbf74a646a/pkg/sql/catalog/catalogkv/catalogkv.go#L285 
</blockquote></details>

I'm going to stick with `UndefinedTable` for consistency with the similar case in `Acquire`.
___
*[pkg/sql/sqlbase/schema_desc.go, line 131 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-M9uxQuqlRXZjSnm_WXx-r11-131:-MBAyKAg-1vZ_5jz75RJ:bmiimwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/sqlbase/schema_desc.go#L131)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This should be a pointer
</blockquote></details>

Done. I didn't even realize not using a pointer receiver still constitutes a valid interface implementation....
___
*[pkg/sql/sqlbase/structured.go, line 1700 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-MBAP6-T2RS-_GtCtJa8:-MBAybp1A1Wdg9sv7PIc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/sqlbase/structured.go#L1700)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

:grimacing: I just realized I had not successfully merged this earlier today. https://github.com/cockroachdb/cockroach/pull/42655 I was planning on bors'ing this, it'll be a minor rebase.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/type_desc.go, line 175 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50236#-MB7mDbNJPND0J4iKjC4-r11-175:-MBAycQMEJbSyuTvkw03:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90fa7195a40cda4b55e3bfc3a03ccb33ae34c0c2/pkg/sql/sqlbase/type_desc.go#L175)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

pointer as well?
</blockquote></details>

Done.

**Comment by lucy-zhang** [comment]
Turns out there was a pre-existing bug in `PublishMultiple`/`Publish` causing no descriptors to be returned, which didn't cause any apparent problems until the advent of the typecasts (in `SchemaChanger.done()`), since we never actually read the result of `done()`. I filed https://github.com/cockroachdb/cockroach/issues/50906 for the bug, but in this PR I just removed the return value.

**Comment by lucy-zhang** [comment]
TFTRs!

bors r+

**Pull request merged**